changes of storeconv.md (#358)
dcs4cop/xcube: docs/source/storeconv.md

## Naming Identifiers

> This section states how identifiers of an xcube data store SHALL look like.

Reviewer: I don't find"of" appropriate here; it misleadingly implies that these identifiers only apply to stores.

"how" is ungrammatical. "How does it look?" and "What does it look like?" are correct sentences, but not "How does it look like?". But in any case I was never entirely happy with the slightly vague use of "look" here. I would suggest something like: "This section defines the structure of the identifiers used by an xcube data store".

Reviewer 2: I agree with Pont.

Reviewer 3: Maybe lose the "the" in Pont's sentence, otherwise I prefer his sentence.

> This section defines the structure of the identifiers used by an xcube data store.

Reviewer 2: ```suggestion
This section explains various identifiers used by the xcube data store framework and defines their format.
```

Within the DataStore framework, identifiers are used to denote data sources, data stores, and data accessors.
Data store, data opener, and data writer identifiers are used to register the component as an extension in a 
package's plugin.py.

> Identifiers MUST be unambiguous within the scope of the data store. 

Reviewer 2: Not really an improvement. I believe original description if fine.

`<type_specifier>` specifies a data type. 
Its intention and format is described in the sub-section below.
In case the type specifier has flags, the flags MUST be given within square brackets, in alphabetic order, separated by 

> single commas and without spaces (e.g., `dataset[cube,multilevel]`).

Reviewer 2: ```suggestion
single commas and without spaces (e.g., `"dataset[cube,multilevel]"`).
```

Type Specifiers are used to specify a data type.
They consist of a name and an arbitrary number of optional flags, given in square brackets. 

> These flags are used to define characteristics of a type, e.g., the type specifier `dataset[cube]` denotes a dataset 

Reviewer 2: ```suggestion
These flags are used to define characteristics of a type, e.g., the type specifier "`dataset[cube]`" denotes a dataset 
```

allowed arguments to `open_data` for each dataset supported by the
`DataOpener` or `DataStore`. The description is provided as a
`JsonObjectSchema` object corresponding to a [JSON
Schema](https://json-schema.org/). The intention is that this description should be 
detailed enough to allow the automatic construction of a user

> interface for accessing the available datasets. Note that, under this system:

Reviewer: I'm not convinced that the rephrasing is an improvement... but perhaps after all it's safest to avoid the subjunctive  in technical documents, so I won't object :).

Reviewer 2: I agree with Pont.

>   1. Every dataset provided by an opener can support a different set of

Reviewer 2: Better!

To work around some of the restrictions of point (2) above, a dataset MAY be
presented by the opener as multiple "virtual" datasets with different
parameter schemas. For instance, the hypothetical dataset described above MAY 

> be offered as two datasets

Reviewer 2: Less clear than the original. Should remain as it was.

`envdata:sensor-a` (with a fixed *sensor_type* of A) and `envdata:sensor-b`,
(with a fixed *sensor_type* of B), offering different sets of permitted

> variables. However, it may not be represented as a single dataset `envdata`.

Reviewer: ```suggestion
variables.
```

No, that's not what I meant! It's also allowed to offer a single dataset. Probably best to remove this clause entirely to avoid misunderstanding.

Sometimes, the interdependencies between parameters are too complex to

> be fully represented by splitting datasets in this manner. In these cases:

Reviewer 2: Ok.

While an opener is free to define any open parameters for any of its datasets,
there are some common parameters which are likely to be used by the majority
of datasets. Furthermore, there are some parameters which are fundamental for the 
description of a dataset and therefore MUST be included in a schema 
(these parameters are denoted explicitly in the list below). In case that an 
opener does not support varying values of one of these parameters, a constant 
value must defined. This may be achieved by the JSON schema's `const` property 

> or by an `enum` property value which is a one-element array.

Reviewer: I find this less clear than the original. If you don't like "whose", I suggest:

> ... or by an `enum` property with a one-element array as its value.

Author: 'Who is always associated with people. Which is used with things. ' https://www.grammarly.com/blog/when-to-use-which-and-who/
That's why I object to using 'whose' in this situation.

Reviewer 2: That is not correct: "whose" does not necessarily refer to a "who" as a person, it may refer to this things as well (de: deren / dessen). See https://www.linguee.de/deutsch-englisch/search?source=auto&query=whose

Therefore please stay with the original.

Reviewer 3: Please maintain word order: "... an enum property WHOSE value ..."

Reviewer: "Who" is not the same as "whose". [Wikipedia has a whole article](https://en.wikipedia.org/wiki/Inanimate_whose) on "inanimate whose", and doesn't list a single modern style guide that recommends against it.

> *Merriam–Webster's Dictionary of English Usage* states that, amongst "the current books" that discussed the subject as of the late 20th century, "not one of them finds [inanimate] whose anything but standard". … It asserts that "notion that whose may not properly be used of anything except persons is a superstition" and such use is "entirely standard as an alternative to of which in all varieties of discourse".

[grammarly.com seems to feel the same way](https://www.grammarly.com/blog/whos-whose/):

> > Bessie carried a lantern, whose light glanced on wet steps and gravel road sodden by a recent thaw.

> This one’s worth an extra glance: “who” in all its forms generally refers to animate beings, but in the possessive there’s no equivalent for inanimate objects like Bessie’s lantern. The very awkward alternative is “Bessie carried a lantern, the light of which glanced on wet steps.” Not good.

Any dataset requiring the specification of these parameters MUST
use the standard parameter names, syntax, and semantics defined below, in
order to keep the interface consistent. For instance, if a dataset allows a
time aggregation period to be specified, it MUST use the `time_period`
parameter with the format described below rather than some other alternative

> name and/or format. Below, the parameters are described with their Python type

Reviewer 2: Ok.

   This parameter MUST be included in an opener parameters schema. 
   
 - `time_range: Tuple[Optional[str], Optional[str]]`  
   The requested time range for the data to be returned. See section
   ‘[Date, time, and duration specifications](#sec-datespec)’.
   This parameter MUST be included in an opener parameters schema.
   
 - `time_period: str`  
   The requested temporal aggregation period for the data. See section
   ‘[Date, time, and duration specifications](#sec-datespec)’.
   This parameter MUST be included in an opener parameters schema.
   
 - `force_cube: bool`  
   Whether to return results as a [specification-compliant
   xcube](https://github.com/dcs4cop/xcube/blob/master/docs/source/cubespec.md).
   If a store supports this parameter and if a dataset is opened with this
   parameter set to `True`, the store MUST return a specification-compliant
   xcube dataset. If this parameter is not supported or if a dataset is opened with
   this parameter set to `False`, the caller MUST NOT assume that the returned
   data conform to the xcube specification.

### Semantics of list-valued parameters

> The `variables` parameter takes as its value a list, with no members duplication

Reviewer: Ungrammatical. "member duplication" (without "s") would be grammatically correct, but I would find "duplicated members" clearer.

Reviewer 3: Or just "duplicate members"

and the values of its members drawn from a predefined set. The values of this

> parameter, and other parameters which values also follow such a format, are

Reviewer: Ungrammatical. The alternative to "whose" would be "... other parameters, the values of which also follow such a format, ...", but I find that rather clunky.

Author: Disagree for the usage of whose, because:  'Who is always associated with people. Which is used with things. ' https://www.grammarly.com/blog/when-to-use-which-and-who/

Reviewer 2: Same as abobe. Pont is right, please leave as-is.

Reviewer: See my answer in previous discussion :).

   corresponding to those members only MUST be returned. In the case of
   `variables`, this will result in a dataset containing only the requested
   variables.

>  - A special case of the above: If an empty list is supplied, a dataset

Reviewer: ```suggestion
 - A special case of the above: if an empty list is supplied, a dataset
```

   containing *no data* MUST be returned -- but with the requested spatial and
   temporal dimensions.

### <a id="sec-datespec"></a>Date, time, and duration specifications

> Within the common parameter `time_range`, times can be specified using the

Reviewer: I'm not sure that this makes things clearer. On consideration, my "can" is also rather vague for a specification document, giving the impression that it's optional. I suggest rewriting this sentence as:

 - In values for the parameter `time_range`, the string components MUST conform to either the `date-time` or `time` JSON schema format.

Reviewer 2: Stay with original.
